src/adapter.rs

```diff
@@ -0,0 +1,630 @@
+use casbin::{Adapter, Model, Result};
+use diesel::{
+    self,
+    r2d2::{ConnectionManager, Pool},
+    result::Error as DieselError,
+    sql_query, BoolExpressionMethods, ExpressionMethods, QueryDsl, RunQueryDsl,
+};
+
+use crate::{error::*, models::*, schema};
+
+use std::error::Error as StdError;
+
+#[cfg(feature = "postgres")]
+use diesel::{pg::PgConnection, PgExpressionMethods};
+
+#[cfg(feature = "postgres")]
+pub struct DieselAdapter {
+    pool: Pool<ConnectionManager<PgConnection>>,
+}
+
+#[cfg(feature = "mysql")]
+use diesel::{dsl::sql, mysql::MysqlConnection};
+#[cfg(feature = "mysql")]
+pub struct DieselAdapter {
+    pool: Pool<ConnectionManager<MysqlConnection>>,
+}
+#[cfg(feature = "mysql")]
+macro_rules! eq_null {
+    ($v:expr,$field:expr) => {{
+        || {
+            use crate::diesel::BoolExpressionMethods;
+
+            sql("")
+                .bind::<diesel::sql_types::Bool, _>($v.is_none())
+                .and($field.is_null())
+                .or(sql("")
+                    .bind::<diesel::sql_types::Bool, _>(!$v.is_none())
+                    .and($field.eq($v)))
+        }
+    }
+    ()};
+}
+
+impl<'a> DieselAdapter {
+    pub fn new(conn_opts: ConnOptions) -> Result<Self> {
+        let manager = ConnectionManager::new(conn_opts.get_url());
+        let pool = Pool::builder().build(manager).map_err(Error::PoolError)?;
+
+        cfg_if! {
+            if #[cfg(feature = "postgres")] {
+                pool.get()
+                    .map_err(|err| Box::new(Error::PoolError(err)) as Box<dyn StdError>)
+                    .and_then(|conn| {
+                        sql_query(format!(
+                            r#"
```

Comment:
I prefer not to create table automatically, sometimes a database account doesn't have permission to do create operation.

Reply:
A common behavior for a Casbin adapter (e.g., all Golang adapters) will try to create a DB named `casbin` if it doesn't exist. If it has no permission, then let it fail.

Reply:
Other adapters like typeorm-adapter (nodejs)  is doing so. 

maybe I should let users choose to create on their own or by this crate.